translated: docs/developer-docs/integrations/t-ecdsa

hokosugi opened the pull request:
## 疑問点・不明点

閾値 ECDSA の暗号技術点に不明点があり四苦八苦といった状況です。
まずはindex.mdからです
* 3行目 「秘密鍵を再構成・復元することなく」と原文にない「復元」を追加しました。これは理解促進のためです。
* 7行目 全体に意訳です。要はECDSA署名のアナロジーとしてHSMに例えているので、伝わればよいと判断してます。

次に、t-ecdsa-how-it-works.md です。
* 3行目 `high-level` を意訳して「俯瞰して」にしました。
* 5行目 同じく `high-level` を「抽象化すると」にしました。
* 7行目 ECDSA  鍵（原文： `ECDSA key` ）に（秘密）を挿入して、「ECDSA （秘密）鍵」にしました。これは、もともと `a new threshold ECDSA key such that the private key is secret shared over the replicas of this subnet` とprivate keyであることを（秘密）を入れることで強調して、理解を早めることを意図してます。これ以後も ECDSA keyと ECDSA マスター鍵の場合（理解が足らず漏れがあるかも）は挿入しています。
* 7行目 `re-sharing` を原文のままにしました。これは一連のchain key技術の中にある再共有の技術名であるためです。後で「再共有」としたところがありますが、これは一般的な用語して使っています。以後のre-sharingは同じ意味合いで使っています。
* 8行目 `quadruple` は、おそらくスカラー倍算の4倍という意味だと思われます。2倍算という言葉はググると出てきますが、４倍算はECDSA署名特有ではないかと思っています。いきなり「４倍算」と出てきてもなんのことか不明なので（スカラー）カッコ付けとしました。
* 11行目 index.mdの３行目と同じく「復元」を挿入。
* 25行目 ロールアウトが何を指しているのか判然としません。恐らく、NNSもAPIを利用しているから安心しろ！っと言いたいと思っていますが、ロールアウトをどう訳せばよいか分からず、そのままにしました。
* 43行目 最後の部分を少し意訳しました。
* 51~55行目 はより専門的なので色々迷います。
      - SEC1 はそのまま、おそらく[このこと](https://www.secg.org/SEC1-Ver-1.0.pdf)
      - `curve` を楕円曲線と訳しました、原文のままがよいでしょうか？（マスタリングBitcoin、マスタリングEthereumともに索引で curve はありません）
     

フォラームで見つけた[Manuのレス](https://forum.dfinity.org/t/threshold-ecdsa-signatures/6152/244)だとECDSA署名は唯一つの秘密鍵をサブネットに保持しているようです。そこから各Canisterではルート鍵が本文にあるタプル「(secp256k1, key_1)」を引数に取る関数で以下無制限に派生鍵を作れる仕組みと思います。ただ、このルート鍵初めとする派生鍵には一時的な「秘密鍵」があるようにも思えるので困るのですが、マスター秘密鍵とCanisterのルート鍵以下は（秘密）を挿入するかしないかで明確に区別したつもりです。
これは自分が秘密鍵かどうかの区別ができず、訳し始めて混乱したためにわかりやすくするつもりで敢えてやっていることです。
そのあたりも考慮してくださると幸甚です。

（追記）
Xnet についてはまだ調べておりません。ググっても固有サービス業者が出てきて、ICに特有な通信の事かもしれません。

hokosugi committed:
translated: docs/developer-docs/integrations/t-ecdsa

hoosan commented on docs/developer-docs/integrations/t-ecdsa/index.md:
```diff
@@ -1,3 +1,26 @@
+# 閾値 ECDSA
+
+Internet Computer は、革新的な閾値 ECDSA プロトコルを実装しています。このプロトコルでは、ECDSA 秘密鍵は秘密分散されて複数のノード、つまり IC 上の閾値 ECDSA 対応サブネットのレプリカによってに保持され、秘密鍵を再構成・復元することなく、分散された秘密鍵を用いて署名されます。このためサブネットの各レプリカは、それ自体では何も情報を提供しない秘密鍵シェアを保持しています。少なくとも3分の1のレプリカがそれぞれの秘密鍵シェアを提供して初めて閾値署名をすることが出来ます。
```
typoの修正です。partiesをノードと訳していただいていますが、parties（当事者）はノードよりも抽象的な用語に思え、少し迷いました。わかりやすさを考え、このままとしました。
復元に関しては、原文にはないことから括弧付けとするのがと良いと思います。
```suggestion
Internet Computer は、革新的な閾値 ECDSA プロトコルを実装しています。このプロトコルでは、ECDSA 秘密鍵は秘密分散されて複数のノード、つまり IC 上の閾値 ECDSA 対応サブネットのレプリカによって保持され、秘密鍵を再構成（復元）することなく、分散された秘密鍵を用いて署名されます。このためサブネットの各レプリカは、それ自体では何も情報を提供しない秘密鍵シェアを保持しています。少なくとも3分の1のレプリカがそれぞれの秘密鍵シェアを提供して初めて閾値署名をすることが出来ます。
```

hoosan commented on docs/developer-docs/integrations/t-ecdsa/index.md:
```diff
@@ -1,3 +1,26 @@
+
+Internet Computer は、革新的な閾値 ECDSA プロトコルを実装しています。このプロトコルでは、ECDSA 秘密鍵は秘密分散されて複数のノード、つまり IC 上の閾値 ECDSA 対応サブネットのレプリカによってに保持され、秘密鍵を再構成・復元することなく、分散された秘密鍵を用いて署名されます。このためサブネットの各レプリカは、それ自体では何も情報を提供しない秘密鍵シェアを保持しています。少なくとも3分の1のレプリカがそれぞれの秘密鍵シェアを提供して初めて閾値署名をすることが出来ます。
+
+Internet Computer の任意のサブネット上の各 Canister は、固有の ECDSA 鍵ペアを制御し、対応する秘密鍵による署名の計算を（ECDSA サブネットに）要求することができます。署名は、その資格のある Canister、すなわち ECDSA 鍵の正当な所持者にのみ発行されます。各 Canister は、固有の ECDSA 鍵に対してのみしか署名を取得することができません。Canister は自分では ECDSA 秘密鍵や秘密鍵シェアを持たず、IC の特定の閾値 ECDSA 対応サブネットに委任していることに注意してください。閾値暗号は、ブロックチェーンのトラストモデルにおいて、従来の暗号だけでは実現不可能な機能を実現するのに役立ちます。
```
has control overは「鍵ペアを制御し」より「鍵ペアの制御を有し」とするのが良いと思いました
```suggestion
Internet Computer の任意のサブネット上の各 Canister は、固有の ECDSA 鍵ペアの制御を有し、対応する秘密鍵による署名の計算を（ECDSA サブネットに）要求することができます。署名は、その資格のある Canister、すなわち ECDSA 鍵の正当な所持者にのみ発行されます。各 Canister は、固有の ECDSA 鍵に対してのみしか署名を取得することができません。Canister は自分では ECDSA 秘密鍵や秘密鍵シェアを持たず、IC の特定の閾値 ECDSA 対応サブネットに委任していることに注意してください。閾値暗号は、ブロックチェーンのトラストモデルにおいて、従来の暗号だけでは実現不可能な機能を実現するのに役立ちます。
```

hoosan commented on docs/developer-docs/integrations/t-ecdsa/index.md:
```diff
@@ -1,3 +1,26 @@
+
+Internet Computer の任意のサブネット上の各 Canister は、固有の ECDSA 鍵ペアを制御し、対応する秘密鍵による署名の計算を（ECDSA サブネットに）要求することができます。署名は、その資格のある Canister、すなわち ECDSA 鍵の正当な所持者にのみ発行されます。各 Canister は、固有の ECDSA 鍵に対してのみしか署名を取得することができません。Canister は自分では ECDSA 秘密鍵や秘密鍵シェアを持たず、IC の特定の閾値 ECDSA 対応サブネットに委任していることに注意してください。閾値暗号は、ブロックチェーンのトラストモデルにおいて、従来の暗号だけでは実現不可能な機能を実現するのに役立ちます。
+
+秘密鍵を安全に保管し、適格なエンティティの要求に応じて、それらに対してのみ署名を発行するハードウェアセキュリティモジュール（HSM）のオンチェーン版としてブロックチェーン上の閾値 ECDSA 実装は見なすことができます。
```
少し言い換えました
```suggestion
ブロックチェーン上の閾値 ECDSA 実装は、秘密鍵を安全に保管し、資格のある対象の要求に応じて（それらに対してのみ）署名を発行するという意味で、ハードウェアセキュリティモジュール（HSM）のオンチェーン版として見なすことができます。
```

hoosan commented on docs/developer-docs/integrations/t-ecdsa/index.md:
```diff
@@ -1,3 +1,26 @@
+閾値 ECDSA が利用できることで、例えば、以下のような多数の重要なユースケースが可能になります：
+-   Bitcoin をネイティブに Canister に保持すること。
+-   Ethereum との統合、例えば Ethereum の ERC-20 トークンを IC に取り込んだり、Ethereum のトランザクションに署名したりすること。
+-   トランザクションへの署名に  ECDSA を署名方式として使用する他のブロックチェーンとの統合すること。
```
冗長さを感じましたので修正しました
```suggestion
-   トランザクションへの署名方式として ECDSA を使用する他のブロックチェーンと統合すること。
```

hoosan commented on docs/developer-docs/integrations/t-ecdsa/t-ecdsa-how-it-works.md:
```diff
@@ -1,3 +1,81 @@
+- 事前署名の計算、署名：これらのプロトコルは秘密分散された（秘密）鍵で署名を計算します。事前署名を計算するための*プロトコル*、すなわち実際の署名プロトコルで使用される（スカラー）4倍算は、事前署名を計算するために署名要求に対して非同期的に実行されます。この事前計算プロトコルは、閾値 ECDSA 署名を作成する手順の大部分を計算します。*署名プロトコル*は、Canister の署名要求によって起動されます。署名プロトコルは、閾値 ECDSA 署名を効率的に計算するために、事前に計算された（スカラー）4倍算を消費します。
+- 公開鍵の取得：Canister が提供する導出パスに基づき Canister の公開鍵を取得することができます。その中には多くの公開鍵導出可能な BIP-32 ライクな鍵導出も含まれます。
+
+秘密鍵は、生成時、サブネット間の秘密鍵シェアの re-sharing 時、署名の計算時など、使用するすべての期間で、秘密鍵シェアされた形でのみ存在し、再構成・復元された形には決してならないことに注意することが重要です。
```
```suggestion
秘密鍵は、生成時、サブネット間の秘密鍵シェアの re-sharing 時、署名の計算時など、使用するすべての期間で、秘密鍵シェアされた形でのみ存在し、再構成（復元）された形には決してならないことに注意することが重要です。
```

hoosan commented on docs/developer-docs/integrations/t-ecdsa/t-ecdsa-how-it-works.md:
```diff
@@ -1,3 +1,81 @@
+IC 上の閾値 ECDSA の概要を俯瞰して説明します。このセクションの情報の一部は、本機能を使用するために必要なものではありませんが、技術的な背景情報を得るために、技術志向の読者にとって興味深いものであるかもしれません。本 IC は、Groth と Shoup の[Eurocrypt 2022 paper](https://eprint.iacr.org/2021/1330) に記載された閾値 ECDSA プロトコルを実装しています。Groth と Shoup は、この分散署名プロトコルの包括的な [design and analysis](https://eprint.iacr.org/2022/506) も発表しています。
+
+IC 上の閾値 ECDSA 実装は、抽象化すると複数のプロトコルを備えています：
+- 鍵生成：指定されたサブネットで実行され、このサブネットのレプリカ上で秘密鍵シェアとして分配される新規の閾値 ECDSA 秘密鍵を生成します。
```
```suggestion
- *鍵生成*：指定されたサブネットで実行され、このサブネットのレプリカ上で秘密鍵シェアとして分配される新規の閾値 ECDSA 秘密鍵を生成します。
```

hoosan commented on docs/developer-docs/integrations/t-ecdsa/t-ecdsa-how-it-works.md:
```diff
@@ -1,3 +1,81 @@
+
+IC 上の閾値 ECDSA 実装は、抽象化すると複数のプロトコルを備えています：
+- 鍵生成：指定されたサブネットで実行され、このサブネットのレプリカ上で秘密鍵シェアとして分配される新規の閾値 ECDSA 秘密鍵を生成します。
+- 鍵の re-sharing（再共有)：このプロトコルはソースサブネットからターゲットサブネットに ECDSA （秘密）鍵を re-sharing します。その結果、ターゲットサブネットのレプリカ上で同じ閾値 ECDSA （秘密）鍵の別のランダムな秘密鍵シェアが使用されて秘密分散されることになります（ソースサブネットが使用する秘密鍵シェアとは可能性として異なる数になり、異なる数のレプリカ上に秘密分散されることになります）。
```
後半、冗長な部分を修正しました
```suggestion
- *鍵の re-sharing（再共有)*：このプロトコルはソースサブネットからターゲットサブネットに ECDSA （秘密）鍵を re-sharing します。その結果、ターゲットサブネットのレプリカ上で同じ閾値 ECDSA （秘密）鍵の別のランダムな秘密鍵シェアが使用されて秘密分散されることになります（ソースサブネットが使用する秘密鍵シェアとは異なる数のレプリカ上に秘密分散される可能性があります）。
```

hoosan commented on docs/developer-docs/integrations/t-ecdsa/t-ecdsa-how-it-works.md:
```diff
@@ -1,3 +1,81 @@
+IC 上の閾値 ECDSA 実装は、抽象化すると複数のプロトコルを備えています：
+- 鍵生成：指定されたサブネットで実行され、このサブネットのレプリカ上で秘密鍵シェアとして分配される新規の閾値 ECDSA 秘密鍵を生成します。
+- 鍵の re-sharing（再共有)：このプロトコルはソースサブネットからターゲットサブネットに ECDSA （秘密）鍵を re-sharing します。その結果、ターゲットサブネットのレプリカ上で同じ閾値 ECDSA （秘密）鍵の別のランダムな秘密鍵シェアが使用されて秘密分散されることになります（ソースサブネットが使用する秘密鍵シェアとは可能性として異なる数になり、異なる数のレプリカ上に秘密分散されることになります）。
+- 事前署名の計算、署名：これらのプロトコルは秘密分散された（秘密）鍵で署名を計算します。事前署名を計算するための*プロトコル*、すなわち実際の署名プロトコルで使用される（スカラー）4倍算は、事前署名を計算するために署名要求に対して非同期的に実行されます。この事前計算プロトコルは、閾値 ECDSA 署名を作成する手順の大部分を計算します。*署名プロトコル*は、Canister の署名要求によって起動されます。署名プロトコルは、閾値 ECDSA 署名を効率的に計算するために、事前に計算された（スカラー）4倍算を消費します。
```
```suggestion
- *事前署名の計算、署名*：これらのプロトコルは秘密分散された（秘密）鍵で署名を計算します。事前署名を計算するための*プロトコル*、すなわち実際の署名プロトコルで使用される（スカラー）4倍算は、事前署名を計算するために署名要求に対して非同期的に実行されます。この事前計算プロトコルは、閾値 ECDSA 署名を作成する手順の大部分を計算します。*署名プロトコル*は、Canister の署名要求によって起動されます。署名プロトコルは、閾値 ECDSA 署名を効率的に計算するために、事前に計算された（スカラー）4倍算を消費します。
```

hoosan commented on docs/developer-docs/integrations/t-ecdsa/t-ecdsa-how-it-works.md:
```diff
@@ -1,3 +1,81 @@
+- 鍵生成：指定されたサブネットで実行され、このサブネットのレプリカ上で秘密鍵シェアとして分配される新規の閾値 ECDSA 秘密鍵を生成します。
+- 鍵の re-sharing（再共有)：このプロトコルはソースサブネットからターゲットサブネットに ECDSA （秘密）鍵を re-sharing します。その結果、ターゲットサブネットのレプリカ上で同じ閾値 ECDSA （秘密）鍵の別のランダムな秘密鍵シェアが使用されて秘密分散されることになります（ソースサブネットが使用する秘密鍵シェアとは可能性として異なる数になり、異なる数のレプリカ上に秘密分散されることになります）。
+- 事前署名の計算、署名：これらのプロトコルは秘密分散された（秘密）鍵で署名を計算します。事前署名を計算するための*プロトコル*、すなわち実際の署名プロトコルで使用される（スカラー）4倍算は、事前署名を計算するために署名要求に対して非同期的に実行されます。この事前計算プロトコルは、閾値 ECDSA 署名を作成する手順の大部分を計算します。*署名プロトコル*は、Canister の署名要求によって起動されます。署名プロトコルは、閾値 ECDSA 署名を効率的に計算するために、事前に計算された（スカラー）4倍算を消費します。
+- 公開鍵の取得：Canister が提供する導出パスに基づき Canister の公開鍵を取得することができます。その中には多くの公開鍵導出可能な BIP-32 ライクな鍵導出も含まれます。
```
BIP-32-like key derivation based on a canister-provided derivation pathとありますので、
Canister が提供する導出パス（canister-provided derivation path）が決定するのはBIP-32-like key derivationだと思います
```suggestion
- *公開鍵の取得*：Canister の公開鍵を取得することができます。その中には Canister が提供する導出パスに基づく BIP-32 ライクな鍵導出も含まれます。
```

hoosan commented on docs/developer-docs/integrations/t-ecdsa/t-ecdsa-how-it-works.md:
```diff
@@ -1,3 +1,81 @@
+
+![Threshold ECDSA Key derivation](../_attachments/key_derivation.png)
+
+閾値 ECDSAのマスター（秘密）鍵は、閾値 ECDSA API では常に*鍵識別子*を通じて参照されます（ロールアウトを管理する NNS のプロポーザルでも同様です）。鍵識別子は楕円曲線名と識別子からなり、このふたつのタプル `(secp256k1, key_1)` は鍵識別子の例です。例えば、署名の計算時に秘密鍵シェアを選択するときや、対応する識別子を持つ鍵を保持する ECDSA 対応サブネットとの間で API コールと応答の XNet ルーティングの実装のときに、これらの鍵識別子は正しい鍵を参照するためにシステムによって使用されます。
```
ここでのロールアウトは、運用開始の管理のことを言っていると思います。
```suggestion
閾値 ECDSAのマスター（秘密）鍵は、閾値 ECDSA API では常に*鍵識別子*を通じて参照されます（運用開始を管理する NNS のプロポーザルでも同様です）。鍵識別子は楕円曲線名と識別子からなり、このふたつのタプル `(secp256k1, key_1)` は鍵識別子の例です。例えば、署名の計算時に秘密鍵シェアを選択するときや、対応する識別子を持つ鍵を保持する ECDSA 対応サブネットとの間で API コールと応答の XNet ルーティングの実装のときに、これらの鍵識別子は正しい鍵を参照するためにシステムによって使用されます。
```

hoosan reviewed (changes requested):
翻訳対応ありがとうございます🙏
疑問点に対しては、コメント適宜参照ください。

> このルート鍵初めとする派生鍵には一時的な「秘密鍵」があるようにも思えるので困る

派生鍵は、元となるルート鍵の位置から少しずらして（tweak）作成しており、秘密鍵はひとつでOKです。ずれてる量が派生鍵ごと異なるので、（単一の）秘密鍵とずれ量（ルート鍵でのずれ＋派生分のずれ）が与えられれば派生鍵の署名ができますが、派生鍵からは秘密鍵の（楕円曲線上における）場所がわからないという関係にあると理解しています

hokosugi committed:
Apply suggestions from code review

ありがとうございました。
全て異論ありません。
この翻訳通じて、秘密鍵の扱いが厳重であることやCanisterが無制限に派生鍵を生成出来ることなど、可能性について学べました。次のユースケース、分散型CAにも魅力感じました。

Co-authored-by: hoosan <40290137+hoosan@users.noreply.github.com>

hoosan committed:
Update docs/developer-docs/integrations/t-ecdsa/index.md
